Allow native crypto properties values YES/NO and TRUE/FALSE

enasser opened the pull request:
The following changes are included in this changeset:
   - Allow the value of native crypto properties as YES/NO in addition
      to TRUE/FALSE as it makes more sense for end user to specify
      jdk.nativeCrypto=NO insted of jdk.nativeCrypto=FALSE.
   - Made the property value case insensitive so that user can
     specify 'YES' or 'Yes' or 'yes'.
   - Allow user to specify enable all native crypto and disable
     specific crypto

Signed-off-by: Nasser Ebrahim <enasser@in.ibm.com>

pshipton commented:
Considering there has been a release (0.11.0) and the user guide updated using the true/false settings, they should continue to be accepted even if yes/no is added as well.

enasser commented:
@pshipton Thank you for your comments. I have modified the code and commit log accordingly.

enasser commented:
Created an issue to document the change in native crypto property: https://github.com/eclipse/openj9-docs/issues/124

pshipton commented on jdk/src/share/classes/com/sun/crypto/provider/CipherCore.java:
```diff
@@ -1227,33 +1225,58 @@ void updateAAD(byte[] src, int offset, int len) {
-            useNativeCBC = Boolean.parseBoolean(nativeCBCStr) || (nativeCBCStr == null);
-            useNativeGCM = Boolean.parseBoolean(nativeGCMStr) || (nativeGCMStr == null);
+            if (nativeCBCStr != null &&
+                (nativeCBCStr.toUpperCase().equals("NO") ||
```
Note that even without this change you can specify NO to disable, since anything but TRUE equates to false. I'm not sure the code should be written to use negative checks, this means you can use anything else such as OFF and its equivalent to TRUE. I think its better to check for YES and TRUE and anything else equates to false.

enasser replied:
Thank you @pshipton for your comments. Currently, any value other than TRUE/FALSE or YES/NO are ignored and it takes the default value which is TRUE/YES. I think that makes more sense than specifying any value with the property change the value from default to FALSE.  Please comment whether you still prefer to reverse the check from NO/FALSE to YES/TRUE in the above change.

enasser commented on jdk/src/share/classes/com/sun/crypto/provider/CipherCore.java:
```diff
@@ -1227,33 +1225,58 @@ void updateAAD(byte[] src, int offset, int len) {
-                return System.getProperty(property);
-            }
+        return AccessController.doPrivileged(
+            (PrivilegedAction<String>) () -> {
```
@pshipton Yes. I noticed that  doPrivileged in CipherCore did not resolve the system property access  issue when security manager is enabled. I just modified it to keep it same as other doPrivileged in this package. I can remove this change as it's not part of the current fix and doPrivileged anyway not working.

pshipton replied:
Up to you, I don't mind if you change it. I suggested the same to Alon in a 2nd revision, but due to time constraints the change was not made.

SueChaplain commented:
@enasser Hi Nasser. Where are we with this work please? I have a doc item open to include the yes|no values as well as the true|false but this PR is still open with discussions.

enasser commented:
@SueChaplain I have addressed the review comments and the PR is ready to review and merge.

@pshipton Can you please review and merge if you approve the fix.

pshipton commented:
@enasser I have reviewed it. As previously stated I don't agree with the changes. I don't think the code should be written to use negative checks, as any input besides NO/FALSE is equivalent to turning on the feature, such as OFF or a typo such as FASLE.

pshipton commented:
@DanHeidinga this is the issue I mentioned, if you have an opinion here.

DanHeidinga commented:
I would prefer the code continue to use `Boolean.parseBoolean` to determine if the property is enabled or not.  Unless there is compelling pre-existing precedent for this kind of change (ie: every other property allows this), we should stick with the original approach of requiring a case-insensitive "true"

enasser commented:
@DanHeidinga Thank you for the review comment. I propose this change as jdk.nativeCrypto=NO makes more sense than jdk.nativeCrypto=FALSE from a user perspective. If you feel jdk.nativeCrypto=false is good enough to disable openssl crypto, then we can discard this PR.

SueChaplain commented:
Wading in from a user perspective. We have a lot of options that are `true|false` where they could also be `yes|no`. Therefore I don't think that the current values will be confusing. What might be more confusing is having 4 possible options to choose from. 

+1 to close the PR. @enasser I think the work you have already done is good enough.
